### PR TITLE
[BEAM-3169] Fixes a data loss bug in WriteFiles when used with fire-once triggers

### PR DESCRIPTION
**Pull request by jkff**
https://issues.apache.org/jira/browse/BEAM-3169

This required a bit of twiddling with shard assignment logic too. The gist of the change is changing the pre-finalize GBK to Reshuffle. I audited all other usages of GBK in the SDK and it appears that only this one is buggy: others either explicitly set a repeated trigger before applying the GBK, or are directly applied to the user's input and the user's trigger firing behavior is WAI.

### CODE REVIEW

**Comment by jkff** [comment]
R: @reuvenlax

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java [hunk 1]
```diff
@@ -563,41 +564,32 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
 
     private int shardNumber;
 
     ApplyShardingKey(
         PCollectionView<Integer> numShardsView,
-        ValueProvider<Integer> numShardsProvider,
         Coder<DestinationT> destinationCoder) {
       this.destinationCoder = destinationCoder;
       this.numShardsView = numShardsView;
-      this.numShardsProvider = numShardsProvider;
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: Please revert these changes as they are not relevant to the fix

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java [hunk 2]
```diff
@@ -682,16 +673,27 @@ public void processElement(ProcessContext context) throws IOException {
       throw new RuntimeException(e);
     }
 
-    if (computeNumShards == null && numShardsProvider == null) {
-      numShardsView = null;
+    final PCollectionView<Integer> fixedNumShardsView;
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: ditto

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java [hunk 3]
```diff
@@ -752,61 +744,79 @@ public void processElement(ProcessContext context) throws IOException {
...
-      keyedResults.setCoder(
-          KvCoder.of(VoidCoder.of(), FileResultCoder.of(shardedWindowCoder, destinationCoder)));
-
-      // Is the continuation trigger sufficient?
+      // We need to materialize the FileResult's before the renaming stage: this can be done either
+      // via a side input or via a GBK. However, when processing streaming windowed writes, results
+      // will arrive multiple times. This means we can't share the below implementation that turns
+      // the results into a side input, as new data arriving into a side input does not trigger the
+      // listening DoFn. We also can't use a GBK because we need only the materialization, but not
+      // the (potentially lossy, dependent on the user's trigger) continuation triggering that GBK
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: I don't think the "potentially-lossy" comment adds much, and is confusing (it's lossy only if the original trigger is lossy).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java [hunk 4]
```diff
@@ -752,61 +744,79 @@ public void processElement(ProcessContext context) throws IOException {
...
+      // would give. So, we use a reshuffle.
       outputFilenames =
-          keyedResults
-              .apply("FinalizeGroupByKey", GroupByKey.<Void, FileResult<DestinationT>>create())
+          results
+              .apply("Reshuffle", Reshuffle.<FileResult<DestinationT>>viaRandomKey())
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: We don't want viaRandomKey, as that will force all bundles to have size 1. We actually want viaConstantKey.

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java [hunk 5]
```diff
@@ -752,61 +744,79 @@ public void processElement(ProcessContext context) throws IOException {
...
+                                window.maxTimestamp(),
+                                window);
                           }
-                          writeOperation.removeTemporaryFiles(tempFiles);
-                          LOG.debug("Removed temporary files for {}.", writeOperation);
+                          LOG.debug(
+                              "Done finalizing write operation for {}.", destEntry.getKey());
                         }
-                      }))
+                        writeOperation.removeTemporaryFiles(tempFiles);
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: I think you dropped the code that adds elements to tempFiles :)

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java (hunk 5 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: Oops. Added a test.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java (hunk 3 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java]: Rephrased.

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by asfgit** [comment]
FAILURE
 
--none--
